File: c++/30804.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

vector<int> v;
int counts[10];

int getTypeCount() {
  int cnt = 0;
  for (int i = 1; i < 10; i++) {
    if (counts[i] != 0) cnt++;
  }

  return cnt;
}

int main() {
  freopen("example.txt", "r", stdin);
  int N;
  cin >> N;

  for (int i = 0; i < N; i++) {
    int a;
    cin >> a;
    v.push_back(a);
  }

  int start = 0;
  int end = 0;
  int ans = 0;

  while (end < N) {
    counts[v[end]]++;

    if (getTypeCount() > 2) {
      counts[v[start]]--;
      start++;
    }

    ans = max(ans, end - start + 1);

    end++;
  }

  cout << ans;

  return 0;
}

File: js/30804.js
const fs = require('fs');

const inputs = fs.readFileSync('example.txt').toString().trim().split('\n');

const N = +inputs[0];
const arr = inputs[1].split(' ').map(Number);

const count = Array(10).fill(0);
let ans = 0;
let start = 0;
let end = 0;

while (end < N) {
  count[arr[end]]++;

  if (getTypeCnt() > 2) {
    count[arr[start++]]--;
  }

  ans = Math.max(ans, end - start + 1);
  end++;
}

console.log(ans);

function getTypeCnt() {
  return count.filter((e) => e !== 0).length;
}
